src/sentry/snuba/metrics/utils.py

```diff
@@ -0,0 +1,143 @@
+__all__ = (
+    "MAX_POINTS",
+    "GRANULARITY",
+    "TS_COL_QUERY",
+    "TS_COL_GROUP",
+    "FIELD_REGEX",
+    "TAG_REGEX",
+    "MetricOperation",
+    "MetricUnit",
+    "MetricType",
+    "OP_TO_SNUBA_FUNCTION",
+    "AVAILABLE_OPERATIONS",
+    "OPERATIONS_TO_ENTITY",
+    "METRIC_TYPE_TO_ENTITY",
+    "ALLOWED_GROUPBY_COLUMNS",
+    "Tag",
+    "TagValue",
+    "MetricMeta",
+    "MetricMetaWithTagKeys",
+    "OPERATIONS",
+    "DEFAULT_AGGREGATES",
+    "UNIT_TO_TYPE",
+    "DerivedMetricParseException",
+    "TimeRange",
+)
+
+
+import re
+from datetime import datetime
+from typing import Collection, Literal, Mapping, Optional, Protocol, Sequence, TypedDict
+
+from sentry.snuba.dataset import EntityKey
+
+MAX_POINTS = 10000
+GRANULARITY = 24 * 60 * 60
+TS_COL_QUERY = "timestamp"
+TS_COL_GROUP = "bucketed_time"
+
+#: Max number of data points per time series:
+# ToDo modify this regex to only support the operations provided
+FIELD_REGEX = re.compile(r"^(\w+)\(((\w|\.|_)+)\)$")
+TAG_REGEX = re.compile(r"^(\w|\.|_)+$")
+
+#: A function that can be applied to a metric
+MetricOperation = Literal["avg", "count", "max", "min", "p50", "p75", "p90", "p95", "p99"]
+MetricUnit = Literal["seconds"]
+#: The type of metric, which determines the snuba entity to query
+MetricType = Literal["counter", "set", "distribution"]
+
+OP_TO_SNUBA_FUNCTION = {
+    "metrics_counters": {"sum": "sumIf"},
+    "metrics_distributions": {
+        "avg": "avgIf",
+        "count": "countIf",
+        "max": "maxIf",
+        "min": "minIf",
+        # TODO: Would be nice to use `quantile(0.50)` (singular) here, but snuba responds with an error
+        "p50": "quantilesIf(0.50)",
+        "p75": "quantilesIf(0.75)",
+        "p90": "quantilesIf(0.90)",
+        "p95": "quantilesIf(0.95)",
+        "p99": "quantilesIf(0.99)",
+    },
+    "metrics_sets": {"count_unique": "uniqIf"},
+}
+
+
+AVAILABLE_OPERATIONS = {
+    type_: sorted(mapping.keys()) for type_, mapping in OP_TO_SNUBA_FUNCTION.items()
+}
+OPERATIONS_TO_ENTITY = {
+    op: entity for entity, operations in AVAILABLE_OPERATIONS.items() for op in operations
+}
+
+METRIC_TYPE_TO_ENTITY: Mapping[MetricType, EntityKey] = {
+    "counter": EntityKey.MetricsCounters,
+    "set": EntityKey.MetricsSets,
+    "distribution": EntityKey.MetricsDistributions,
+}
+
+ALLOWED_GROUPBY_COLUMNS = ("project_id",)
+
+
+class Tag(TypedDict):
+    key: str  # Called key here to be consistent with JS type
+
+
+class TagValue(TypedDict):
+    key: str
+    value: str
+
+
+class MetricMeta(TypedDict):
+    name: str
+    type: MetricType
+    operations: Collection[MetricOperation]
+    unit: Optional[MetricUnit]
+
+
+class MetricMetaWithTagKeys(MetricMeta):
+    tags: Sequence[Tag]
+
+
+_OPERATIONS_PERCENTILES = (
+    "p50",
+    "p75",
+    "p90",
+    "p95",
+    "p99",
+)
+
+OPERATIONS = (
+    "avg",
+    "count_unique",
+    "count",
+    "max",
+    "sum",
+) + _OPERATIONS_PERCENTILES
```

Comment:
Q: Is there a way to deduplicate this with the `MetricOperation` type?

Reply:
Not sure what this is in reference to? can you elaborate?

Reply:
Near the top of the file there is a typing declaration for the `MetricOperation` literal type which lists effectively the same operation names as `_OPERATIONS_PERCENTILES`. I was wondering if there's a way to deduplicate this so that you only have a single list of those operation names.